### ISSUE TITLE
Request: convenience functions vector.above and vector.below

### ISSUE THREAD

**Issue by kab0u** [minetest/minetest#13171]
A very common pattern for lua code is to check nodes above or below a given map position.  While this can be expressed effectively using vector.offset(pos, 0, 1, 0) and vector.offset(pos, 0, -1, 0) it would be nice to have a shorthand alternative.

The availabilty of vector functions vector.above(pos) and vector.below(pos) would positively benefit code readability and also save mod authors a small amount of typing.

**Comment by sorcerykid** [comment]
Minetest S3 engine has `minetest.get_node_above(pos, y)`, which saves the intermediary step of creating another vector for such a commonly used function. To get the node below, you just pass a negative `y` value.

It should be easy enough to port over, so I will probably open a PR.

**Comment by kab0u** [comment]
Yes, that would also be useful.

**Comment by TurkeyMcMac** [comment]
`minetest.get_node_above(pos, y)` will not be very advantageous once https://github.com/minetest/minetest/pull/13017 is merged, since LuaJIT will be able to optimize out the extra vector.

**Comment by rubenwardy** [comment]
Above/below is confusing when the same terms are used for pointed things

Would be better to use something like:

```lua
pos + vector.up
```

**Comment by sorcerykid** [comment]
> minetest.get_node_above(pos, y) will not be very advantageous once https://github.com/minetest/minetest/pull/13017 is merged, since LuaJIT will be able to optimize out the extra vector.

It's not a question of speed, but rather code readability. Nesting function calls for a very common operation is something I personally try to avoid.

**Comment by SmallJoker** [comment]
:-1: The Lua API should be kept as generic as possible and not cluttered with nice-to-have utility functions. If a mod needs this vector operation a lot, a local function might make more sense for their use-case.

**Comment by appgurueu** [comment]
Agreed with SmallJoker.

There is no real issue here; `pos:offset(0, 1, 0)` is hardly inconvenient or unreadable.

First of all, consider the incompleteness of the proposed API: Why should `above` and `below` be special in any way? The other 4 directions also ought to be considered.

Second, the naming: `above` and `below` might create confusion as pointed out by ruben. Having this "utils" in the `vector` namespace is indeed "cluttering" and plain dirty, esp. if you had all 4 directions (left, right, back, front).

At this point, a cleaner API would be to just store the 6 directions - up, down, left, right, front, back - but that would be confusing with X vs Z and cluttering the vector namespace. You could use a subtable `vector.dir` for it, but then you lose the small benefit of fewer keystrokes.

Finally, all of this can always be left up to mods to do locally to adhere to whatever convention they consider cleanest. Library mods can help abstract it away.

**Comment by ancientmarinerdev** [comment]
> Above/below is confusing when the same terms are used for pointed things
> 
> Would be better to use something like:
> 
> ```lua
> pos + vector.up
> ```

Isn't the problem with that that above and below for pointed_thing are badly named?

    * `under` refers to the node position behind the pointed face.
    * `above` refers to the node position in front of the pointed face.

I cannot understand how those names passed review. It should be in front and behind, or closest, furthest. You should always design things how they are supposed to be, rather than how they have been, and move towards it.

"in or to a [higher](https://dictionary.cambridge.org/dictionary/english/higher) [position](https://dictionary.cambridge.org/dictionary/english/position) than something [else](https://dictionary.cambridge.org/dictionary/english/else):

The dictionary definition of above, is higher, and that isn't a relative concept. It is absolute.

I would personally advocate for pointed_thing deprecate above and below for more meaningfully accurate names like pointed_thing.behind and pointed_thing.in_front and give meaningful options for vector such as .up and .down, or .above and .below.

If you have to read documentation to understand something, it's probably not the best name. If left, right, front, back are ambigious (which they are), it shouldn't prevent those that are explicit like above and below.

On the topic of clutter, an API is supposed to be used, and if the decision not to clutter the API is focussed on a maintenance perspective and not the user of that API, it is missing a key part. API's are only as good as their usage and if you have to clutter to use, the clutter is still there, it's just pass to the client code.

**Comment by kab0u** [comment]
Thank you all for responding and adding valuable insights.

@rubenwardy indeed, "up" and "down" are much better choices, both for the sake of brevity and for avoiding confusion with existing nomenclature.  Thanks for the suggestion.

@ancientmarinerdev you have a point, and "front" and "behind" (or "inside" and "outside") may be better fitting terms.  However that is in hindsight and the terms are more or less cast in stone.  A case can be made to change this and keep above and below as legacy aliases, but I doubt if it is a high priority for most minetest developers.

@appgurueu the common offset cases are "above"/"up" (y+1), "below"/"down" (y-1) and "flat" (y=0).  offsets in the x- and z-directions are much less common patterns.

@sorcerykid you are correct that my primary intent with the functions is code readability.  An extra layer of function calls only happens if the function relies on `vector.offset()` for its implementation.  If instead it implements an analogue of `vector.offset()`, there is no performance loss.

In summary, a case can be made both for and against inclusion and I understand it if that is not a priority to most or even undesirable to some minetest developers. 

In any case, for future reference to mod developers, I can attest that the following code works well when placed in a mod and has a good potential for making code look more readable.

```lua
function vector.up(v, n)
	if type(n) ~= "number" then n = 1 end
	return setmetatable({x = v.x, y = v.y + n, z = v.z}, vector.metatable)
end
```

**Comment by kab0u** [comment]
My description above of `vector.flat()` may have been lacking a bit of clarity, so I'll try to improve on that with its definition:
```lua
-- Returns the projection of vector v onto the x-z plane.
function vector.flat(v)
	return setmetatable({x = v.x, y = 0, z = v.z}, vector.metatable)
end
```
And some more convenient vector extensions:

```lua
-- Returns a vector n nodes below vector v, or one node if n is not specified.
function vector.down(v, n)
	if type(n) ~= "number" then n = 1 end
	return setmetatable({x = v.x, y = v.y - n, z = v.z}, vector.metatable)
end

-- Returns true is vector v is inside the box bounded by vectors a and b.
function vector.in_area(v, a, b)
	local p, q = vector.sort(a, b)
	return p.x <= v.x and v.x <= q.x and p.y <= v.y and v.y <= q.y and p.z <= v.z and v.z <= q.z
end

-- Returns true if vector v is inside radius r around point p.
function vector.in_radius(v, p, r)
	return vector.distance(v, p) <= r
end
```
